flight/Libraries/math/smoothcontrol.c

```diff
@@ -0,0 +1,195 @@
+			axis->differential = interp_diff / (float)axis->control_interval;
+			axis->integrator_timeout = axis->control_interval;
+			break;
+
+		// Linear interpolation of predicts of the type above.
+		case SMOOTHCONTROL_SUPERSMOOTH:;
+			float target = axis->signal + signal_diff *
+				(SMOOTHCONTROL_SMOOTH_CHAMFER_START+
+					SMOOTHCONTROL_SMOOTH_PREDICTOR_SLOPE*
+					SMOOTHCONTROL_SMOOTH_DUTY_CYCLE);
+			axis->differential = (target - axis->current) / (float)axis->control_interval;
+			axis->integrator_timeout = axis->control_interval;
+			break;
+	}
+
+	axis->signal = new_signal;
+}
+
+// Resets the axis state.
+void smoothcontrol_reinit(smoothcontrol_state state, uint8_t axis_num, float new_signal)
+{
+	ASSERT(!state);
+
+	// Prevent integrating when stick scale changes.
+	state->axis[axis_num].differential = 0;
+	state->axis[axis_num].signal = new_signal;
+	state->axis[axis_num].current = new_signal;
+	// Once the minimum interval was found, it shouldn't change.
+	// Re-initialize anyway. Too large of an assumed interval will cause integration to be
+	// too slow on proper timing, anyway.
+	state->axis[axis_num].control_interval = 255;
+}
+
+// Sets mode for an axis.
+void smoothcontrol_set_mode(smoothcontrol_state state, uint8_t axis_num, uint8_t mode)
+{
+	ASSERT(!state || axis_num > 3);
+	state->axis[axis_num].mode = mode;
+	smoothcontrol_reinit(state, axis_num, state->axis[axis_num].signal);
+}
+
+// Processes the signal, if internal state allows it.
+void smoothcontrol_run(smoothcontrol_state state, uint8_t axis_num, float *new_signal)
+{
+	ASSERT(!state || axis_num > 3);
+
+	struct smoothcontrol_axis_state *axis = &state->axis[axis_num];
+
+	// No mode, bypass.
+	if(axis->mode == SMOOTHCONTROL_NONE) return;
+
+	if(*new_signal != axis->signal) {
+		// Ignore anything at 1 tick and below. Only happens when stabilization loop
+		// starts and throws off proper interval detection.
+		if(axis->tick_counter > 1 && axis->tick_counter < axis->control_interval) {
+			axis->control_interval = axis->tick_counter;
+		}
+
+		// Signal changed, set up integrator on the new stuff.
+		smoothcontrol_update(state, axis, *new_signal);
+
+		axis->tick_counter = 0;
+	}
+
+	if(axis->tick_counter <= axis->integrator_timeout && axis->tick_counter < state->time_bomb) {
+		// Don't fiddle with first value after update.
+		if(axis->tick_counter > 0) axis->current += axis->differential;
+	} else {
+		// Kill the differential after integrator time-out, otherwise things may wind up.
+		axis->differential = 0;
+
+		// If we passed the timebomb, force the receiver signal, too.
+		if(axis->tick_counter >= state->time_bomb) axis->current = *new_signal;
+	}
+
+	*new_signal = axis->current;
+}
+
+// Separate handling of thrust.
+void smoothcontrol_run_thrust(smoothcontrol_state state, float *new_signal)
+{
+	// The -1 on zero throttle causes a huge blip when throttling up, so we need to
+	// handle it separately.
+	if(new_signal < 0)
+	{
+		float tmp = 0;
```

Comment:
really really need to think about the safety of this, especially with fancy interpolators.  Can a negative throttle signal ever be interpolated to positive?